File: app/src/main/java/be/digitalia/fosdem/viewmodels/TracksListViewModel.java
package be.digitalia.fosdem.viewmodels;

import android.app.Application;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.arch.core.util.Function;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import be.digitalia.fosdem.db.AppDatabase;
import be.digitalia.fosdem.model.Day;
import be.digitalia.fosdem.model.Track;

public class TracksListViewModel extends AndroidViewModel {

	private final AppDatabase appDatabase = AppDatabase.getInstance(getApplication());
	private final MutableLiveData<Day> day = new MutableLiveData<>();
	private final LiveData<List<Track>> tracks = Transformations.switchMap(day,
			new Function<Day, LiveData<List<Track>>>() {
				@Override
				public LiveData<List<Track>> apply(Day day) {
					return appDatabase.getScheduleDao().getTracks(day);
				}
			});

	public TracksListViewModel(@NonNull Application application) {
		super(application);
	}

	public void setDay(@NonNull Day day) {
		if (!day.equals(this.day.getValue())) {
			this.day.setValue(day);
		}
	}

	public LiveData<List<Track>> getTracks() {
		return tracks;
	}
}

File: app/src/main/kotlin/be/digitalia/fosdem/viewmodels/TracksListViewModel.kt
package be.digitalia.fosdem.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import be.digitalia.fosdem.db.ScheduleDao
import be.digitalia.fosdem.flow.stateFlow
import be.digitalia.fosdem.flow.versionedResourceFlow
import be.digitalia.fosdem.inject.CallbackViewModelAssistedFactory
import be.digitalia.fosdem.model.Day
import be.digitalia.fosdem.model.Track
import dev.zacsweers.metro.AppScope
import dev.zacsweers.metro.Assisted
import dev.zacsweers.metro.AssistedFactory
import dev.zacsweers.metro.AssistedInject
import dev.zacsweers.metro.ContributesIntoMap
import dev.zacsweers.metro.binding
import dev.zacsweers.metrox.viewmodel.ViewModelAssistedFactory
import dev.zacsweers.metrox.viewmodel.ViewModelAssistedFactoryKey
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.filterNotNull

@AssistedInject
class TracksListViewModel(
    @Assisted day: Day,
    scheduleDao: ScheduleDao,
) : ViewModel() {

    val tracks: Flow<List<Track>> = stateFlow(viewModelScope, null) {
        versionedResourceFlow(scheduleDao.version) {
            scheduleDao.getTracks(day)
        }
    }.filterNotNull()

    @AssistedFactory
    @ContributesIntoMap(AppScope::class, binding = binding<ViewModelAssistedFactory>())
    @ViewModelAssistedFactoryKey(TracksListViewModel::class)
    fun interface Factory : CallbackViewModelAssistedFactory {
        fun create(day: Day): TracksListViewModel
    }
}
